 Data Cleaning - to do 

- Clean and standardize column names.
- Verify data types.
- Remove `UDI`.
- Extract `L`, `M`, `H` from `Product ID`.
- Remove the serial number.
- Remove/exclude `TWF`, `HDF`, `PWF`, `OSF`, `RNF` from ML features.
- Check invalid values and ranges.
- Recheck missing values and duplicates.


In [4]:
# libraries

import pandas as pd
import numpy as np

In [5]:
df = pd.read_csv("../data/raw/ai4i2020.csv")

In [6]:
df.head(5)

,UDI,Product ID,Type,Air temperature [K],Process temperature [K],Rotational speed [rpm],Torque [Nm],Tool wear [min],Machine failure,TWF,HDF,PWF,OSF,RNF
0,1,M14860,M,298.1,308.6,1551,42.8,0,0,0,0,0,0,0
1,2,L47181,L,298.2,308.7,1408,46.3,3,0,0,0,0,0,0
2,3,L47182,L,298.1,308.5,1498,49.4,5,0,0,0,0,0,0
3,4,L47183,L,298.2,308.6,1433,39.5,7,0,0,0,0,0,0
4,5,L47184,L,298.2,308.7,1408,40.0,9,0,0,0,0,0,0


In [7]:
print(df.shape)

(10000, 14)


In [10]:
print("missing value",df.isnull().sum().sum() )
print()
print("duplicate values", df.duplicated().sum())

missing value 0

duplicate values 0


In [12]:
# removing the unwanted columns 

df.drop(columns = ['UDI','Product ID', 'TWF','HDF','PWF','OSF','RNF'], inplace = True)

In [13]:
df.shape

(10000, 7)

In [16]:
for i in df.columns:
    print(i)

Type
Air temperature [K]
Process temperature [K]
Rotational speed [rpm]
Torque [Nm]
Tool wear [min]
Machine failure


In [21]:
# Cleaning column names

df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
    .str.replace("[", "", regex=False)
    .str.replace("]", "", regex=False)
)

In [22]:
df.columns

Index(['type', 'air_temperature_k', 'process_temperature_k',
       'rotational_speed_rpm', 'torque_nm', 'tool_wear_min',
       'machine_failure'],
      dtype='str')

In [23]:
df.dtypes

type                         str
air_temperature_k        float64
process_temperature_k    float64
rotational_speed_rpm       int64
torque_nm                float64
tool_wear_min              int64
machine_failure            int64
dtype: object

In [26]:
# data types 
df["type"] = df["type"].astype("category")

In [29]:
df.dtypes

type                     category
air_temperature_k         float64
process_temperature_k     float64
rotational_speed_rpm        int64
torque_nm                 float64
tool_wear_min               int64
machine_failure             int64
dtype: object

### checking invalid range and values . Speed should not be in negative..things like that 

In [33]:
df.describe()

,air_temperature_k,process_temperature_k,rotational_speed_rpm,torque_nm,tool_wear_min,machine_failure
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000,10000.000000
mean,300.004930,310.005560,1538.776100,39.986910,107.951000,0.033900
std,2.000259,1.483734,179.284096,9.968934,63.654147,0.180981
min,295.300000,305.700000,1168.000000,3.800000,0.000000,0.000000
25%,298.300000,308.800000,1423.000000,33.200000,53.000000,0.000000
50%,300.100000,310.100000,1503.000000,40.100000,108.000000,0.000000
75%,301.500000,311.100000,1612.000000,46.800000,162.000000,0.000000
max,304.500000,313.800000,2886.000000,76.600000,253.000000,1.000000


## Range and Value Check conclusion 

- Air temperature ranges from 295.3 K to 304.5 K.
- Process temperature ranges from 305.7 K to 313.8 K.
- Rotational speed ranges from 1168 rpm to 2886 rpm.
- Torque ranges from 3.8 Nm to 76.6 Nm with no negative values.
- Tool wear ranges from 0 to 253 minutes.
- Machine failure contains only valid binary values (0 and 1).
- No immediately invalid or impossible values were identified.
- Some extreme values may exist, particularly in rotational speed and torque, but they should be investigated during EDA rather than removed during data cleaning.

In [34]:
# overall data cleaning is done . Will save the df into data/processed as "cleaned_df.csv"

In [40]:
df.to_csv("../data/processed/cleaned_df.csv", index=False)